pandas/_typing.py

```diff
@@ -26,8 +26,11 @@
 )
 ArrayLike = TypeVar("ArrayLike", ABCExtensionArray, np.ndarray)
 DatetimeLikeScalar = TypeVar("DatetimeLikeScalar", Period, Timestamp, Timedelta)
-Dtype = Union[str, np.dtype, ExtensionDtype]
+Dtype = Union[str, np.dtype, "dtypes.ExtensionDtype"]
 FilePathOrBuffer = Union[str, Path, IO[AnyStr]]
 
 FrameOrSeries = TypeVar("FrameOrSeries", ABCSeries, ABCDataFrame)
 Scalar = Union[str, int, float]
+
+# TODO(GH26403): Replace with Optional[bool] or bool
+OrderedType = Union[None, bool, object]
```

Comment:
Always go for `Optional` instead of `Union[None...]`. 

Also `object` is very rarely what you want in Typing so would want to use `Any` instead, though the `Union` of `Any` and another object also doesn't make sense since `Any` encompasses everything...

Can't this just be `Optional[bool]`?

Reply:
Also reading through the entire diff I see you just moved this from elsewhere so no worries but makes sense to correct here

Reply:
``Optional`` only works with a single argument, so we have to use Union here. I think the need for object.

The object type will also be removed in 1.0, so this type can just be ``bool``. I think @TomAugspurger will deal with that. If you look at the comment in core/dtypes/dtypes.py, line 21 to see the reasoning.

Reply:
Gotcha but the sentinel never actually gets returned from `def ordered` below right? If that's correct then I think should be `Ordered = Optional[bool]` and in the __init__ for CategoricalDtype should be `Union[Ordered, object]`

Reply:
Unfortunately I can't get that to work. ``Categorical.ordered``'s type is same as ``CategoricalDtype._ordered`` which typewise may be a object (but actually can't be).

I don't have a good solution, except waiting for ``ordered_sentinel`` to be removed in 1.0?

Reply:
Can you share the error it gives you? If the problem is that the `ordered_sentinel` is in the constructor of `CategoricalDtype` I think could just explicitly annotate that parameter to be `Optional[bool]` as well, which would be cleanest in the long run

Reply:
``pandas\core\arrays\categorical.py:475: error: Incompatible return value type (got "object", expected "Optional[bool]")``

It all goes back to the ``ordered`` parameter being of type ``Union[None, bool, object]``, while the ``ordered`` attribute should be of type Optional[bool]`` (but isn't).

Don't know if there's a pattern for having different parameter and attribute types...

Reply:
Can you just add `# type: ignore` to that one line? I think that is the lesser of all evils and easiest to revert whenever that change gets made

Reply:
Very nitpicky but can you also change the name from `OrderedType`? That would seem to imply that this is something like a TypeVar or NewType. 

Either `OrderedAlias` or just plain `Ordered` would be a better fit